documentation/monads.md

```diff
@@ -0,0 +1,292 @@
+Console.WriteLine(
+    (from i in new Lazy<int>(() => 42)
+    from s in new Lazy<string>(i.ToString)
+    select $"{i}:{s}").Value); // output: "42:42"
+```
+
+### Tasks
+
+Tasks are another built in type of monad that has direct compiler support with ``async/await``.
+
+By providing the appropriate ``SelectMany`` implementations for LINQ it is also possible to further increase
+readability[^1].
+
+So instead of writing:
+
+```c#
+return await serviceB.DoSomethingAsync(
+    (await serviceA.LoadExecutorAsync()).ExecuteAsync()
+)
+```
+
+we could write[^2]
+
+```c#
+return from executor in serviceA.LoadExecutorAsync()
+from executionResult in executor.ExecuteAsync()
+select serviceB.DoSomethingAsync(executionResult)
+```
+
+This not only eliminates all ``await`` calls but also parentheses that were needed previously.
+
+[^1]: whether this actually improves readability is debatable and/or depends on the reader
+[^2]: Granted, we could have used a less compact version to begin with, like
+      ```c# 
+      var executor = await serviceA.LoadExecutorAsync(); 
+      var executionResult = await executor.ExecuteAsync(); 
+      return await serviceB.DoSomethingAsync(executionResult)
+      ```
+      But this is not shorter than the LINQ version and not relevant for the purpose of showing the monadic nature of the Task
+      type.
+
+## What is this library about?
+
+This library provides some common monad types that are not part of the standard dotnet library.
+
+### Explicit signatures
+
+Consider the following aptly named method.
+
+```c#
+public static string? Foo(){
+    return null;
+}
+```
+
+This method returns a nullable string, but we have no idea what this means. Is null a regular value, or is this used to
+indicate an error?
+
+By i.e. using a ``MayBe`` monad instead we can ensure the caller, yes it is normal to get nothing from this method, and
+this is not a sign that something went wrong.
+
+```c#
+public static Maybe<string> Foo(){
+    return Nothing;
+}
+```
+
+If we use the ``Either`` monad instead we can explicitly say normally a non null string is returned, but there might be
+error cases, that need to be handled separately.
+
+```c#
+public static Either<string,Error> Foo(){
+    return Error.NotFound.Left();
+}
+```
+
+One step further we can use a ``Maybe`` inside of an ``Either`` to signal that there is a possibility to get a string,
+no value at all, or an error from this method[^3].
+
+```c#
+public static Either<Maybe<string>,Error> Foo(){
+    return Error.NotFound.Left();
+}
+```
+
+The same can be done for parameters as well.
+All in all this minimizes the possibility of ``lying`` signatures.
+
+[^3]: However, you should not overdo this, because, like all generic types, stacking monads will get extremely
+      unreadable very quickly. To solve this problem we'd need direct language support
+      for [Discriminated/Tagged Unions](https://en.wikipedia.org/wiki/Tagged_union) but even then only up to a certain point.
+
```

Comment:
ich fände einen abschnitt über 'how to unit test monads' sehr hilfreich
(vielleicht auch im vergleich - mit und ohne monad)